application/btsieve/src/ethereum/queries/event.rs

```diff
@@ -114,49 +113,94 @@ impl QueryType for EventQuery {
-    // in the receipt. This needs to be fixed with https://github.com/comit-network/comit-rs/issues/638
-    fn expand_result(
-        result: &QueryResult,
-        client: Arc<Web3<Http>>,
+    type Item = PayloadKind;
+
+    fn to_http_payload(
+        &self,
+        return_as: &ReturnAs,
+        client: &Web3<Http>,
     ) -> Result<Vec<Self::Item>, Error> {
-        let futures: Vec<_> = result
-            .0
-            .iter()
-            .filter_map(|tx_id| match hex::decode(clean_0x(tx_id)) {
-                Ok(bytes) => Some(bytes),
-                Err(e) => {
-                    warn!("Skipping {} because it is not valid hex: {:?}", tx_id, e);
-                    None
-                }
-            })
-            .map(|id| {
-                client
-                    .eth()
-                    .transaction(TransactionId::Hash(H256::from_slice(id.as_ref())))
-                    .map_err(Error::Web3)
-            })
-            .collect();
+        let to_payload = |transaction_id: H256| to_payload(client, transaction_id, return_as);
 
-        stream::futures_ordered(futures)
+        self.0
+            .iter()
+            .filter_map(to_h256)
+            .map(to_payload)
+            .collect::<FuturesOrdered<_>>()
             .filter_map(|item| item)
             .collect()
             .wait()
     }
 }
 
+fn to_payload(
+    client: &Web3<Http>,
+    transaction_id: H256,
+    return_as: &ReturnAs,
+) -> Box<dyn Future<Item = Option<PayloadKind>, Error = Error>> {
+    let transaction_future = client
+        .eth()
+        .transaction(TransactionId::Hash(transaction_id))
+        .map(|maybe_transaction| maybe_transaction.map(Box::new))
+        .map_err(Error::Web3);
+
+    let receipt_future = client
+        .eth()
+        .transaction_receipt(transaction_id)
+        .map(|maybe_receipt| maybe_receipt.map(Box::new))
+        .map_err(Error::Web3);
+
+    match return_as {
+        ReturnAs::Transaction => Box::new(transaction_future.map(|maybe_transaction| {
+            maybe_transaction.map(|transaction| PayloadKind::Transaction { transaction })
+        })),
+        ReturnAs::Receipt => Box::new(
+            receipt_future
+                .map(|maybe_receipt| maybe_receipt.map(|receipt| PayloadKind::Receipt { receipt })),
+        ),
+        ReturnAs::TransactionId => {
+            Box::new(future::ok(Some(PayloadKind::Id { id: transaction_id })))
+        }
+        ReturnAs::TransactionAndReceipt => Box::new(transaction_future.join(receipt_future).map(
+            move |maybe| match maybe {
+                (Some(transaction), Some(receipt)) => Some(PayloadKind::TransactionAndReceipt {
+                    transaction,
+                    receipt,
+                }),
+                // TODO: Investigate under which circumstances the ethereum node fails to return a
+                // transaction/receipt
+                (None, _) => {
+                    warn!(
```

Comment:
:question: Why do you log the absence of a return element in the `ReturnAs::TransactionAndReceipt` match arm and not for the other ones?

Reply:
Good question :D 
Should probably log it for all of them :)